### PR TITLE
Implements group-based authorization

### LINKED ISSUE #69: limit to group/role?
I'm migrating from an LDAP realm to an SSO realm, and I'm looking at both keycloak and gluu. For both OpenID IdP, the notion of limiting a user's access to a specific client/SP by group (or role) is argued to be a topic of *authorization* (vice OIDC's *authentication*), and therefore the responsibility of the SP.

Is it possible to limit the use of NC by the presence of groups or roles in the user's attributes?

### PR DESCRIPTION
**Pull request by djessich**
Support for group-based authroization is implemented in this commit. Only
the configured group defined in plugin configuration is able to access
Nextcloud. This authorized group is being read from optional
'oidc_login_authorized_group' configuration property.

In case the property is configured in plugin configuration and the user
is not assigned to the configured group, the login will not be allowed
for the user by failing the login process with an appropriate error
message. The user groups will be read from OIDC response using the
attribute configured in 'oidc_login_attributes' configuration property.
When the user is not authorized, the user will neither be created nor
its data will be updated in Nextcloud.

Fixes #69

### CODE REVIEW

**Comment by pulsejet** [comment]
This is the job of the OIDC provider

**Comment by chris246** [comment]
Unfortunately, not all providers support this, as e.g. Keycloak

**Comment by sirkrypt0** [comment]
> Unfortunately, not all providers support this, as e.g. Keycloak

Hey,

Keycloak does not support this feature natively, but it can be added via a plugin. Have a look at [my comment](https://github.com/pulsejet/nextcloud-oidc-login/issues/69#issuecomment-926181354) on the issue you referenced.

**Comment by djessich** [comment]
@pulsejet On linked issue #69 you commented:
> Ideally you should do this on the provider side. But I'll be happy to review any PRs to implement this too.
(see https://github.com/pulsejet/nextcloud-oidc-login/issues/69#issuecomment-776547156)

So I implemented this feature as I think it would be a good improvement to the plugin. With that improvement one can decide to either use the provider or this plugin to authorize a user group.

**Comment by Ornias1993** [comment]
@djessich please rebase so @pulsejet can finally have a look...

**Comment by djessich** [comment]
@Ornias1993 and @ykuksenko Yes, I'll do that asap.

**Comment by djessich** [comment]
Sorry for my late reply, the PR is now rebased. @pulsejet Could you check? Thanks in advance!

**Comment by djessich** [comment]
I also fixed some style errors reported by CI build in the code I changed.

**Comment by djessich** [comment]
I also included a fix for the rest of issues reported during PHP-CS-fixer build. Its not my code, but to have this PR ready to be merged. Waiting for maintainer to approve the build.

File: README.md
Comment (azmeuk) [inline README.md:116]: The example is a single group, but your code supports array of groups. I suggest either:
- make the example configuration an array
- support both single values and arrays

**Comment by djessich** [comment]
I updated the style fixes with latest result from CI.

File: README.md
Comment (djessich) [reply README.md:116]: Acutally the my new check does not support an array of groups (LoginController.php#213 and below). I just moved the original code a few lines higher to ensure that no user/group is created before my new check applies. This is why it is just required to check, that a single authorized group (as configured) is among the groups retrieved from the OIDC provider. So, I only verify if the configured group (not an array) is among those groups.

Should I really support multiple groups and allow configuration for an array and a single group? I thought it is enough to restrict it to just a single group, as you can put all users accessing Nextcloud that should be authorized into this new group.

Comment (djessich) [reply README.md:116]: @azmeuk Please a short response how we should continue. Thanks!

Comment (azmeuk) [reply README.md:116]: I think you are right. Let's start with support for a single group, and if someone someday needs support for multiple group, we can think again.

**Review by pulsejet — CHANGES_REQUESTED** [review]

**Pull request closed without merging**
